docs/design.md

```diff
@@ -0,0 +1,106 @@
+by image creator. Meanwhile, this approach can facilitate the development of standardization of
+implementing Enclave-CC, which reuses the methodology of image protections already applied to the
+container ecosystem, such as image [encryption](https://github.com/containers/ocicrypt-rs) and
+[signing](https://github.com/containers/image/blob/main/docs/containers-signature.5.md), instead
+of using LibOSes specific approaches.
+
+
+![no_wrapping.png](docs/images/no_wrapping.png)
+
+### Align with Confidential Containers principles
+
+Confidential Containers design principle tends to utilize and reuse ecosystem to implement
+Confidential Containers. More specifically, Confidential Containers requires:
+
+- Container image is always protected, e.g, signed and/or encrypted.
+- Download and unpack container image to the secure and attested enclave for the actual workloads
+to run with the protections of confidentiality and integrity.
+- Prove that sensitive workloads are running on a genuine and trusted hardware TEE.
+- Follow up trust and threat model defined by Confidential Containers.
+
+Enclave-CC approach would strictly align with above design principles to implement a general
+process-based Confidential Containers co-existing with VM-based Confidential Containers. It
+provides a consistent Kubernetes user experience across hardware TEEs, but also compared with the
+existing, non confidential Kubernetes usage patterns. The container workload is unmodified (but
+encrypted/signed) and identical across hardware TEEs.
+
+![k8s_ux.png](docs/images/k8s_ux.png)
+
+where:
+- [Inclavare Containers Project](https://github.com/inclavare-containers/inclavare-containers) is
+a container runtime to launch the process-based Confidential Container for Intel SGX.
+
+- LibOS generally refers to the SGX LibOS implementations, such as
+[Gramine](https://github.com/gramineproject/gramine) and
+[Occlum](https://github.com/occlum/occlum).
+
+- `rune` is the container runtime for spawning and running enclaves in containers.
+
+- `runelet` is the containerized process used to interact with LibOS through common PAL APIs.
+
+## Architecture
+
+![arch.png](docs/images/arch.png)
+
+enclave-CC introduces a helper enclave called `agent enclave` which is a fixed function enclave.
+It is launched by `shim-rune` during pod creation in a private method. More specifically,
+`agent enclave` is deployed in the form of OCI bundle rather than a container image, and thus
+`shim-rune` can call `rune` to start it. `agent enclave` has the same life cycle as the pod.
+Its main job is to receive and then load the actual workload's container image into an
+`app enclave`.
+
+To do this securely both the agent and app enclaves establish trust through attestation.
+
+`agent enclave` is responsible for hosting a LibOS loaded from OCI bundle previously mentioned
+and `enclave-agent` on top of the LibOS. `enclave-agent` consists of
```

Comment:
In order to be compatible with the process model of Gramine, the two services will be combined together to one process on top of libos.

Reply:
I was going to bring up this as an open too. I started thinking yesterday do we want to "combine" them together or just run them as separate ones. AFAICS `attestation-agent` does not necessarily have to be per-pod but it could run on the node. This would require secure gRPC though.

Reply:
I think hosting them together would be easy to be managed. `agent enclave` as a whole can be managed and provisioned. If AA is hosted in another enclave, it would increase the communication complexity.

Reply:
Both options have pros and cons and taking fresh look for what works best for enclave-cc is the best thing to do. A totally different approach could also be PKCS11 for which there are some building blocks available.